src/ucs/time/timerq.c

```diff
@@ -85,37 +95,75 @@ ucs_status_t ucs_timerq_add(ucs_timer_queue_t *timerq, int timer_id,
     return status;
 }
 
-ucs_status_t ucs_timerq_remove(ucs_timer_queue_t *timerq, int timer_id)
+ucs_status_t ucs_timerq_remove(ucs_timer_queue_t *timerq, uint64_t timer_id,
+                               ucs_timer_t *hint)
 {
     ucs_status_t status;
     ucs_timer_t *ptr;
 
-    ucs_trace_func("timerq=%p timer_id=%d", timerq, timer_id);
+    ucs_trace_func("timerq=%p timer_id=%lu", timerq, timer_id);
 
     status = UCS_ERR_NO_ELEM;
 
     ucs_spin_lock(&timerq->lock);
+    if (hint && ucs_likely(hint->id == timer_id)) {
+        ucs_assert(hint >= timerq->timers);
+        ucs_assert(hint < timerq->timers + timerq->used_cnt);
+        *hint = timerq->timers[--timerq->used_cnt];
```

Comment:
what if the location '*hint' becomes invalid because multiple timers before it were removed?

Reply:
@yosefe  right you are. I should have made the hint an (unsigned) offset...

Reply:
this is still a problem , is something is changed inside the array the offset which user code has also becomes out of date

Reply:
@yosefe I agree, but in my use-case timers are usually added and removed shortly after (timeout for fault-tolerance), so it is likely that the location will remain the same. I agree this is not guaranteed, so I called it a "hint" - indicating it's only "best-effort".

Reply:
having both timer_id and a hint seems redundant. Is it possible to have O(1) remove just by timer_id? for example like we do in ucs_callbackq ?

Reply:
@yosefe If I understand correctly, what you are proposing is changing timerq to have a ID->idx mapping array for fast access. Looks like this implies a relatively big change:

1. Timer ID has to be generated by timerq. Otherwise selecting ID=(unsigned)-1 is a problem), and we don't want collisions.
2. This means TimerQ becomes a "growing array with ID->idx translation", which is very similar to CallbackQ (and ptr_array) - makes sense to rewrite them to use a common datastruct (only one "grow" function").
3. if we change callbackQ's ID selection - this goes through `ucs_async_signal_timerq_add_timer` and `ucs_async_signal_add_timer`, which will change the `ucs_async_ops_t` API... I don't mind writing the change - but it will effects a relatively large number of places.

Is this what you had in mind?

Reply:
in general yes, but thinking of it more - if there are many timers in the list, the dispatch will also be slow because it has to go over all of them, even if most of the timers are not expired, right?

Reply:
@yosefe The dispatch will indeed be slow if there are many timers, but this is not related to this (or proposed) patch. This change is just about accelerating removal/modification by adding a location hint.

The main way to address slowness due to a large number of timers is to have more Timer Queues. Making each TimerQueue service only one interval was something I've been thinking about for my fault-tolerance feature, but maybe an overkill at this stage.

I'm assuming you're not concerned about the overhead of ID->idx (or idx->ID in case of removal), because those should be O(1) (though more instructions than today).

Reply:
I see that timerq is used only by "async" today. maybe change ucs_async_add_timer() function so instead of passing id to timerq, it will let timerq allocate its own id (which will be the array offset) and save this id in async handler, and then use it for removal?
this means timerq will generate the id according to a free slot in the array

Reply:
@yosefe yeah, that's what I figured - just wanted to make sure it's ok before implementing. I'll take care of it this week. Also, I think I'll close this PR, because the new code will be very different...

Reply:
@alex--m i would suggest splitting to 2 PRs?
1. in ucs/async - keep additional timer id inside the handler, as if it's coming from timerq
2. in timerq  - generate id instead of accepting id